### PR TITLE
[MPS] Fall back multi-layer LSTM on macOS 12

### LINKED ISSUE #90421: LSTM for MPS backend is broken
### 🐛 Describe the bug

A bidirectional LSTM using the MPS backend gives bad results, with or without batch_first, regardless of the number of layers.

There was an existing bug report which addressed one aspect of this problem, but it was closed before the problem was fully resolved.  There are still several cases in which an LSTM on MPS is broken.

https://github.com/pytorch/pytorch/issues/80306

Here is a short script which demonstrates the issue:

```
import torch

def test_lstm(num_layers, bidirectional, batch_first):
    torch.manual_seed(1234)
    lstm = torch.nn.LSTM(5, 5, num_layers=num_layers, bidirectional=bidirectional, batch_first=batch_first)
    lstm.eval()
    inp = torch.randn(4, 1, 5)
    if batch_first:
        inp = inp.transpose(0, 1)
    print("num_layers {} bidirectional {} batch_first {}".format(num_layers, bidirectional, batch_first), end="  ")
    print(torch.linalg.norm(lstm(inp)[0]).item(), end="  ")
    print(torch.linalg.norm(lstm.to("mps")(inp.to("mps"))[0]).item())

test_lstm(2, True, True)   # bad
test_lstm(2, True, False)  # bad
test_lstm(1, True, True)   # bad

test_lstm(2, False, True)  # ok
test_lstm(1, False, True)  # ok
test_lstm(1, False, False) # ok
```

The results:

```
/Users/john/stanza/bar.py:12: UserWarning: The operator 'aten::linalg_vector_norm' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  print(torch.linalg.norm(lstm.to("mps")(inp.to("mps"))[0]).item())
num_layers 2 bidirectional True batch_first True  0.9131234288215637  2.432673692703247
num_layers 2 bidirectional True batch_first False  0.9131234288215637  2.432673692703247
num_layers 1 bidirectional True batch_first True  0.8858160972595215  2.428898334503174
num_layers 2 bidirectional False batch_first True  0.5294497013092041  0.5294497609138489
num_layers 1 bidirectional False batch_first True  0.6045955419540405  0.6045956611633301
num_layers 1 bidirectional False batch_first False  0.6045955419540405  0.6045956611633301
```

### Versions

Environment information:
```
Collecting environment information...
PyTorch version: 1.14.0.dev20221206
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.6 (default, Sep 26 2022, 11:37:49)  [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.14.0.dev20221206
[pip3] torchaudio==0.14.0.dev20221206
[pip3] torchvision==0.15.0.dev20221206
[conda] Could not collect
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @kulinseth @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by qqaatw**
The native implementation of LSTM has been fixed on macOS 13.

On macOS 12, the multi-layer LSTM still has a numerical correctness issue that cannot be resolved on OS's side.

Thus, we fall back the multi-layer LSTM on macOS 12 to LSTMCell iteration. It might have performance impact but will make LSTM on macOS 12 fully usable.

Fixes: #90421
Issues related: #80306, #83144

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### CODE REVIEW

**Comment by kulinseth** [comment]
@qqaatw , can you please rebase this PR

**Comment by qqaatw** [comment]
> @qqaatw , can you please rebase this PR

Done.

File: test/test_mps.py [hunk 1]
```diff
@@ -1704,25 +1704,28 @@ def helper(operator):
             if operator == "<=":
                 res_mps = x_mps <= y_mps
                 res_cpu = x_cpu <= y_cpu
-            if operator == "<":
+            elif operator == "<":
```
Comment (albanD) [inline test/test_mps.py:1707]: This overlaps with the other PR right?

File: test/test_mps.py [hunk 2]
```diff
@@ -7981,65 +7984,73 @@ def test_cpu_indices(self, device="mps"):
...
-    @unittest.skipIf(True, "Backward of lstm returns wrong result")
-    def test_lstm_2(self, device="mps", dtype=torch.float32):
-        def get_results(device):
-            rnn = nn.LSTM(1, 4, 1, device=device)
-            inp = torch.randn(2, 3, 1, device=device, requires_grad=True)
-            hx = torch.zeros(1, 3, 4, device=device)
-            cx = torch.zeros(1, 3, 4, device=device)
-
-            output, _ = rnn(inp, (hx, cx))
+    def test_lstm_forward(self, device="mps", dtype=torch.float32):
```
Comment (albanD) [inline test/test_mps.py:7987]: This should really be tested via ModuleInfo. Is there a PR for that already @kulinseth ?

File: aten/src/ATen/native/mps/operations/RnnOps.mm [hunk 3]
```diff
@@ -393,14 +419,29 @@
 
 
                         gradientTensor_ = [outputs objectAtIndex:0];
-                        [gradOutputArray addObject:[outputs objectAtIndex:0]];
+                        MPSGraphTensor* gradientOutputTensor_ = [outputs objectAtIndex:0];
+                        if (batch_first) {
+                            gradientOutputTensor_ = [mpsGraph transposeTensor:gradientOutputTensor_
+                                                        dimension:0
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:425]: indentation looks bad?

**Review by albanD — APPROVED** [review]
That sounds good to me.
Will let @kulinseth take a final look.

One-off testing is not the best though and we should transition out of it!

File: aten/src/ATen/DeviceGuard.h [hunk 4]
```diff
@@ -24,7 +24,7 @@ inline c10::optional<Device> device_of(const Tensor& t) {
 }
 
 inline c10::optional<Device> device_of(const c10::optional<Tensor>& t) {
-  return t.has_value() ? device_of(t.value()) : nullopt;
+  return t.has_value() ? device_of(t.value()) : c10::nullopt;
```
Comment (qqaatw) [inline aten/src/ATen/DeviceGuard.h:27]: The `nullopt` conflicts with the one in the standard library of macOS, where I don't remember which specifically...., so makes it unambiguous.

File: aten/src/ATen/mps/MPSDevice.h [hunk 5]
```diff
@@ -1,9 +1,9 @@
 //  Copyright © 2022 Apple Inc.
 
 #pragma once
+#include <c10/core/Allocator.h>
 #include <c10/macros/Macros.h>
 #include <c10/util/Exception.h>
-#include <ATen/ATen.h>
```
Comment (qqaatw) [inline aten/src/ATen/mps/MPSDevice.h:6]: Avoid include `ATen.h` here due to the `#define TORCH_ASSERT_ONLY_METHOD_OPERATORS` assertion in `RNN.cpp`.

File: test/test_mps.py (hunk 2 above)
Comment (qqaatw) [reply test/test_mps.py]: cc @kulinseth.

**Review by kulinseth — APPROVED** [review]

File: test/test_mps.py (hunk 2 above)
Comment (kulinseth) [reply test/test_mps.py]: Yeah there is Module info PR, agreed, we can change it as part [of that PR](https://github.com/pytorch/pytorch/pull/95334).

**Pull request closed without merging**